File: frontend/src/services/proposalEvaluationService.ts
import { api } from "./api";

/* ============================================================
   TYPES
============================================================ */

export type ProposalEvaluationDecision =
  | "UNDER_REVIEW"
  | "SHORTLISTED"
  | "REJECTED";

export interface UniversityProposal {
  id: number;

  rfp_id: number;
  invitation_id: number;
  university_id: number;
  submitted_by: number;

  title: string;
  solution: string;
  technical_approach: string;

  research_methodology: string | null;
  required_resources: string | null;
  estimated_cost: number | null;
  expected_timeline_days: number | null;
  faculty_team: string | null;
  expected_outcomes: string | null;
  technology_requirements: string | null;

  status:
    | "DRAFT"
    | "SUBMITTED"
    | "UNDER_EVALUATION"
    | "SHORTLISTED"
    | "REJECTED"
    | "WITHDRAWN";

  submitted_at: string | null;
  created_at: string;
  updated_at: string;
}


export interface ProposalEvaluation {
  id: number;
  proposal_id: number;
  evaluated_by: number;

  technical_feasibility_score: number;
  innovation_score: number;
  cost_effectiveness_score: number;
  impact_score: number;
  timeline_score: number;
  scalability_score: number;
  research_capability_score: number;

  overall_score: number;

  remarks: string | null;
  decision: ProposalEvaluationDecision;

  evaluated_at: string;
  created_at: string;
  updated_at: string;
}


export interface CreateProposalEvaluationRequest {
  proposal_id: number;

  technical_feasibility_score: number;
  innovation_score: number;
  cost_effectiveness_score: number;
  impact_score: number;
  timeline_score: number;
  scalability_score: number;
  research_capability_score: number;

  remarks?: string;
}


export interface ProposalEvaluationDecisionRequest {
  decision: "SHORTLISTED" | "REJECTED";
  remarks?: string;
}


/* ============================================================
   GET PROPOSALS FOR RFP
============================================================ */

export async function getRFPProposals(
  rfpId: number
): Promise<UniversityProposal[]> {
  const response = await api.get<UniversityProposal[]>(
    `/university-proposals/rfp/${rfpId}`
  );

  return response.data;
}


/* ============================================================
   GET SINGLE PROPOSAL
============================================================ */

export async function getUniversityProposal(
  proposalId: number
): Promise<UniversityProposal> {
  const response = await api.get<UniversityProposal>(
    `/university-proposals/${proposalId}`
  );

  return response.data;
}


/* ============================================================
   CREATE EVALUATION
============================================================ */

export async function createProposalEvaluation(
  payload: CreateProposalEvaluationRequest
): Promise<ProposalEvaluation> {
  const response = await api.post<ProposalEvaluation>(
    "/proposal-evaluations",
    payload
  );

  return response.data;
}


/* ============================================================
   DECIDE PROPOSAL
============================================================ */

export async function decideProposalEvaluation(
  evaluationId: number,
  payload: ProposalEvaluationDecisionRequest
): Promise<ProposalEvaluation> {
  const response = await api.post<ProposalEvaluation>(
    `/proposal-evaluations/${evaluationId}/decision`,
    payload
  );

  return response.data;
}

File: mediator_backend/app/services/proposal_evaluation_service.py
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.proposal_evaluation import (
    ProposalEvaluation,
    ProposalEvaluationDecision,
)
from app.models.university_proposal import (
    UniversityProposal,
    UniversityProposalStatus,
)
from app.models.user import User, UserRole
from app.schemas.proposal_evaluation import ProposalEvaluationCreate


WEIGHTS = {
    "technical_feasibility_score": 0.20,
    "innovation_score": 0.15,
    "cost_effectiveness_score": 0.10,
    "impact_score": 0.20,
    "timeline_score": 0.10,
    "scalability_score": 0.10,
    "research_capability_score": 0.15,
}


def calculate_overall_score(
    evaluation: ProposalEvaluationCreate,
) -> float:
    score = (
        evaluation.technical_feasibility_score
        * WEIGHTS["technical_feasibility_score"]
        + evaluation.innovation_score
        * WEIGHTS["innovation_score"]
        + evaluation.cost_effectiveness_score
        * WEIGHTS["cost_effectiveness_score"]
        + evaluation.impact_score
        * WEIGHTS["impact_score"]
        + evaluation.timeline_score
        * WEIGHTS["timeline_score"]
        + evaluation.scalability_score
        * WEIGHTS["scalability_score"]
        + evaluation.research_capability_score
        * WEIGHTS["research_capability_score"]
    )

    return round(score, 2)


def evaluate_proposal(
    db: Session,
    evaluation_data: ProposalEvaluationCreate,
    current_user: User,
) -> ProposalEvaluation:

    if current_user.role not in {
        UserRole.SUPER_ADMIN,
        UserRole.GOVERNMENT_OFFICER,
        UserRole.MUNICIPALITY_OFFICER,
    }:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only authorized government officers can evaluate proposals.",
        )

    proposal = (
        db.query(UniversityProposal)
        .filter(
            UniversityProposal.id == evaluation_data.proposal_id
        )
        .first()
    )

    if not proposal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="University proposal not found.",
        )

    if proposal.status != UniversityProposalStatus.SUBMITTED:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Only submitted proposals can be evaluated.",
        )

    existing = (
        db.query(ProposalEvaluation)
        .filter(
            ProposalEvaluation.proposal_id
            == evaluation_data.proposal_id
        )
        .first()
    )

    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="This proposal has already been evaluated.",
        )

    overall_score = calculate_overall_score(
        evaluation_data
    )

    # Initial decision threshold.
    # Government can later override this through a separate
    # shortlist/rejection workflow.
    decision = ProposalEvaluationDecision.UNDER_REVIEW

    evaluation = ProposalEvaluation(
        proposal_id=proposal.id,
        evaluated_by=current_user.id,

        technical_feasibility_score=(
            evaluation_data.technical_feasibility_score
        ),
        innovation_score=evaluation_data.innovation_score,
        cost_effectiveness_score=(
            evaluation_data.cost_effectiveness_score
        ),
        impact_score=evaluation_data.impact_score,
        timeline_score=evaluation_data.timeline_score,
        scalability_score=evaluation_data.scalability_score,
        research_capability_score=(
            evaluation_data.research_capability_score
        ),

        overall_score=overall_score,
        remarks=evaluation_data.remarks,
        decision=decision,
    )

    proposal.status = UniversityProposalStatus.UNDER_EVALUATION

    db.add(evaluation)
    db.commit()
    db.refresh(evaluation)

    return evaluation

def decide_proposal(
    db: Session,
    evaluation_id: int,
    decision: str,
    remarks: str | None,
    current_user: User,
) -> ProposalEvaluation:

    if current_user.role not in {
        UserRole.SUPER_ADMIN,
        UserRole.GOVERNMENT_OFFICER,
        UserRole.MUNICIPALITY_OFFICER,
    }:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only authorized government officers can decide proposal outcomes.",
        )

    evaluation = (
        db.query(ProposalEvaluation)
        .filter(
            ProposalEvaluation.id == evaluation_id
        )
        .first()
    )

    if not evaluation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Proposal evaluation not found.",
        )

    proposal = evaluation.proposal

    if decision not in {
        ProposalEvaluationDecision.SHORTLISTED.value,
        ProposalEvaluationDecision.REJECTED.value,
    }:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Decision must be SHORTLISTED or REJECTED.",
        )

    evaluation.decision = ProposalEvaluationDecision(
        decision
    )

    if remarks is not None:
        evaluation.remarks = remarks

    if decision == ProposalEvaluationDecision.SHORTLISTED.value:
        proposal.status = UniversityProposalStatus.SHORTLISTED

    elif decision == ProposalEvaluationDecision.REJECTED.value:
        proposal.status = UniversityProposalStatus.REJECTED

    db.commit()
    db.refresh(evaluation)

    return evaluation